DDInstall/DDInstall.psm1

```diff
@@ -0,0 +1,215 @@
+
+    .PARAMETER Hostname
+    Configures the hostname reported by the Agent to Datadog (overrides any hostname calculated at runtime).
+
+    .PARAMETER DDAgentUsername
+    Override the default ddagentuser username used during Agent installation (v6.11.0+).
+
+    .PARAMETER DDAgentPassword
+    Override the cryptographically secure password generated for the ddagentuser user during Agent installation (v6.11.0+). Must be provided for installs on domain servers/domain controllers.
+
+    .PARAMETER ApplicationDataDirectory
+    Override the directory to use for the configuration file directory tree (v6.11.0+). May only be provided on initial install; not valid for upgrades. Default: C:\ProgramData\Datadog
+
+    .PARAMETER ProjectLocation
+    Override the directory to use for the binary file directory tree (v6.11.0+). May only be provided on initial install; not valid for upgrades. Default: %ProgramFiles%\Datadog\Datadog Agent
+
+    .PARAMETER AgentInstallURL
+    Override the URL which the Agent installer is downloaded from.
+
+    .PARAMETER InstallLogPath
+    Override the Agent installation log location.
+
+    .INPUTS
+    None. You cannot pipe objects to Install-DDAgent.
+
+    .OUTPUTS
+    None.
+
+    .EXAMPLE
+    C:\PS> Install-DDAgent -ApiKey "<YOUR_DATADOG_API_KEY>"
+
+    .EXAMPLE
+    C:\PS> Install-DDAgent -ApiKey "<YOUR_DATADOG_API_KEY>" -WithAPMTracers "DotNet"
+
+    .LINK
+    Learn more about the Datadog Windows Agent User:
+    https://docs.datadoghq.com/agent/guide/windows-agent-ddagent-user/
+    #>
+}
+
+###############################
+# Unexported helper functions #
+###############################
+
+function formatAgentInstallerParameters($params)
+{
+    $formattedparams = "APIKEY=$($params.ApiKey)"
+
+    if ($params.ContainsKey('Site'))                        { $formattedparams += " SITE=$($params.Site)"}
+    if ($params.ContainsKey('Tags'))                        { $formattedparams += " TAGS=``$($params.Tags)``" }
+    if ($params.ContainsKey('Hostname'))                    { $formattedparams += " HOSTNAME=$($params.Hostname)" }
+    if ($params.ContainsKey('DDAgentUsername'))             { $formattedparams += " DDAGENTUSER_NAME=$($params.DDAgentUsername)" }
+    if ($params.ContainsKey('DDAgentPassword'))             { $formattedparams += " DDAGENTUSER_PASSWORD=$($params.DDAgentPassword)" }
+    if ($params.ContainsKey('ApplicationDataDirectory'))    { $formattedparams += " APPLICATIONDATADIRECTORY=$($params.ApplicationDataDirectory)" }
+    if ($params.ContainsKey('ProjectLocation'))             { $formattedparams += " PROJECTLOCATION=$($params.ProjectLocation)" }
+
+    return $formattedparams
+}
+
+function downloadAsset($url, $outFile)
+{
+    # Suppress progress bar to prevent performance issues (affects older windows versions, see https://github.com/PowerShell/PowerShell/issues/13414)
+    $ProgressPreference = 'SilentlyContinue'
+
+    Invoke-WebRequest -UseBasicParsing -Uri $url -OutFile $outFile
```

Comment:
You could try
`(New-Object System.Net.WebClient).DownloadFile($url, $outFile)`

It gets rid of the progress banner (so you might be able to get rid of the `SilentlyContinue`.  It's also way faster.

Reply:
Also, with this I believe you want to set a fully qualified target path ($outFile).

Reply:
^ is a good call - in a previous comment, I said I felt strongly about the progress bar because otherwise it took > 5 minutes of looking at what appeared to be a hung prompt. If we want to do away with the progress bar, then I think at a minimum we need to print "Download In Progress" or something, so we know that it is actually doing something.

I still feel strongly that we need some indication that something is happening / the script isn't broken, but I'm open to it not being a progress bar. Maybe like a `[=]` that spins so like: `[=]`  --> `[||]` -> `[=]`.

Reply:
(Yes, that looks horrible, but I don't know how else to represent it in a static comment 😅 )

Reply:
since `DownloadFile` is a blocking call, I think you're stuck w/o active progress.
  
  You could certainly add `Write_host` checkpoints along the way, i.e.
  ```
  Beginning download file ...
  ... done download
  Beginning install...
  ... done install```

Reply:
This is the current output:
```
> Install-DDAgent -WithAPMTracers DotNet -ApiKey xxxxx
Downloading Datadog Windows Agent installer
Installing Datadog Windows Agent
Downloading Datadog .NET Tracing Library v2.49.0
Installing .NET Tracing Library
```
Do we think that's sufficient?